quinn-proto v0.10.0 w/ rustls v0.21.0

cpu opened the pull request:
# Description 

Update branch for quinn-proto v0.10.0, using Rustls v0.21.0

### deps: upgrade rustls v0.20.3 -> v0.21.0.
This commit updates Quinn's dependency on Rustls from v0.20.3 to v0.21.0.

This release includes a handful of small breaking changes which are resolved alongside the upgrade.

### quinn-proto: bump version 0.9.2 -> 0.10.0.

As part of upgrading the quinn-proto rustls dependency we must do a semver incompatible version bump since rustls is exposed as part of the public API.

### Blocking tasks

- [x] Review
- [x] Wait for upstream Rustls v0.21.0 release
- [x] Replace patch with update to use released tag
- [x] Rebase on main

cpu commented on quinn-proto/src/crypto/rustls.rs:
```diff
@@ -326,7 +323,12 @@ fn to_vec(params: &TransportParameters) -> Vec<u8> {
 
 pub(crate) fn initial_keys(version: Version, dst_cid: &ConnectionId, side: Side) -> Keys {
-    let keys = rustls::quic::Keys::initial(version, dst_cid, side.is_client());
+    let side = if side.is_client() {
+        rustls::Side::Client
+    } else {
+        rustls::Side::Server
+    };
```
This was a quick fix. I think probably it would be nicer to add a `From<Side> for rustls::Side` impl to convert between the Quinn `Side` and the Rustls `Side` :thinking:

cpu replied:
Tidied up in [bf15998](https://github.com/quinn-rs/quinn/pull/1515/commits/bf159983708465eaa80ba7d5ca2745bf33ad3d50).

cpu commented on quinn-proto/src/tests/mod.rs:
```diff
@@ -352,9 +352,10 @@ fn reject_self_signed_server_cert() {
     let client_ch = pair.begin_connect(client_config_with_certs(vec![]));
     pair.drive();
-    assert_matches!(pair.client_conn_mut(client_ch).poll(),
+    let res = pair.client_conn_mut(client_ch).poll();
+    assert_matches!(res,
```
I pulled the `res` out of `assert_matches!` while debugging. Probably it could be reverted, but I mildly prefer this to make it easier to breakpoint and examine

djc reviewed (commented):
Thanks for starting on this!

Would be nice to squash most of these changes (except the one that seems like a technically unrelated bugfix), and also treat this branch/PR as the eventual upgrade PR (removing some of the test language from the PR title/description).

Also would be nice to include a separate commit that bumps the version number for quinn-proto to 0.10.0 (since we expose rustls publicly as part of our API, we need to do a semver-incompatible bump here if we make a semver-incompatible bump for rustls). The quinn crate already got one in #1502, so it doesn't need to be bumped.

cpu commented on quinn-proto/src/crypto/rustls.rs:
```diff
@@ -180,7 +179,8 @@ impl crypto::Session for TlsSession {
     ) -> Result<(), ExportKeyingMaterialError> {
         self.inner
             .export_keying_material(output, label, Some(context))
-            .map_err(|_| ExportKeyingMaterialError)
```
It's required in order to have the code build with Rustls v0.21.0-alpha. I think it should be included with the other changes in order to keep each commit building/passing tests.

The upstream `export_keying_material` return type changed from `Result((), ...)` to `Result<T, ...>` in [38fdd952bec1ca9ab7c3b25b60026954115561a9](https://github.com/rustls/rustls/commit/38fdd952bec1ca9ab7c3b25b60026954115561a9) which means this code can't rely on the `Ok(())` being passed through `map_err` anymore.

cpu replied:
@djc I think doing this might be a little tricky and I want to make sure you think it's worthwhile before charging ahead.

To match the upstream exactly we'll have to add a generic bound `T: <AsMut<[u8]>>` to the `export_keying_material` function defined in the Quinn `Session` trait that the rustls `TlsSession` implements. Doing that breaks the object safety of the `Session` trait, and there are ~8 places where existing Quinn code is boxing `Box<dyn Session>`.

WDYT?

djc replied:
Ahhh yeah, let's not do that.

cpu commented:
> Would be nice to squash most of these changes (except the one that seems like a technically unrelated bugfix), and also treat this branch/PR as the eventual upgrade PR (removing some of the test language from the PR title/description).

:ballot_box_with_check: - Done.

> Also would be nice to include a separate commit that bumps the version number for quinn-proto to 0.10.0

:ballot_box_with_check: - Done w/ 0ad878c

Ralith reviewed (dismissed):
Thanks!

cpu commented:
> cpu force-pushed the cpu-smoke-test-rustls-quic-conns branch from 0ad878c to e17beaa 

I wanted to refresh this branch now that https://github.com/rustls/rustls/pull/1216 was merged. I'm glad I did because it shook out a couple errors, now addressed in https://github.com/rustls/rustls/pull/1236 I've switched the `Patch` to that branch for the short term.

cpu commented:
> cpu force-pushed the cpu-smoke-test-rustls-quic-conns branch from e17beaa to 429380a 

Fixed one more breaking change from upstream:
```
error[E0432]: unresolved import `rustls::internal::msgs::enums::AlertDescription`
  --> quinn-proto/src/tests/mod.rs:13:5
   |
13 | use rustls::internal::msgs::enums::AlertDescription;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `AlertDescription` in `msgs::enums`
```

cpu commented:
> cpu force-pushed the cpu-smoke-test-rustls-quic-conns branch from 429380a to d79ed33 

Now back to tracking `rustls/main` in the `Patch` since https://github.com/rustls/rustls/pull/1236 was merged.

cpu commented:
@djc @Ralith This branch should be ready to go now that Rustls 0.21.0 is available :tada:

djc reviewed (approved):
Thanks!

Ralith reviewed (approved):
Thanks!
